javascript/ql/src/semmle/javascript/frameworks/Vue.qll

```diff
@@ -0,0 +1,359 @@
+    ) {
+      extend.hasLocationInfo(filepath, startline, startcolumn, endline, endcolumn)
+    }
+
+    override DataFlow::Node getOwnOption(string name) { result = extend.getOptionArgument(0, name) }
+
+  }
+
+  /**
+   * An instance of an extended Vue, for example `instance` of `var Ext = Vue.extend({...}); var instance = new Ext({...})`.
+   */
+  class ExtendedInstance extends Instance, MkExtendedInstance {
+    VueExtend extend;
+
+    DataFlow::NewNode sub;
+
+    ExtendedInstance() { this = MkExtendedInstance(extend, sub) }
+
+    override string toString() { result = sub.toString() }
+
+    override predicate hasLocationInfo(
+      string filepath, int startline, int startcolumn, int endline, int endcolumn
+    ) {
+      sub.hasLocationInfo(filepath, startline, startcolumn, endline, endcolumn)
+    }
+
+    override DataFlow::Node getOwnOption(string name) { result = sub.getOptionArgument(0, name) }
+
+    override DataFlow::Node getOption(string name) {
+      result = Instance.super.getOption(name)
+      or
+      result = MkExtendedVue(extend).(ExtendedVue).getOption(name)
+    }
+
+  }
+
+  /**
+   * A Vue component from `Vue.component("my-component", { ... })`.
+   */
+  class Component extends Instance, MkComponent {
+    DataFlow::CallNode def;
+
+    Component() { this = MkComponent(def) }
+
+    override string toString() { result = def.toString() }
+
+    override predicate hasLocationInfo(
+      string filepath, int startline, int startcolumn, int endline, int endcolumn
+    ) {
+      def.hasLocationInfo(filepath, startline, startcolumn, endline, endcolumn)
+    }
+
+    override DataFlow::Node getOwnOption(string name) { result = def.getOptionArgument(1, name) }
+
+  }
+
+  /**
+   * A single file Vue component in a `.vue` file.
+   */
+  class SingleFileComponent extends Instance, MkSingleFileComponent {
+    VueFile file;
+
+    SingleFileComponent() { this = MkSingleFileComponent(file) }
+
+    override predicate hasLocationInfo(
+      string filepath, int startline, int startcolumn, int endline, int endcolumn
+    ) {
+      filepath = file.getAbsolutePath() and
+      startline = 0 and
+      startcolumn = 0 and
+      endline = 0 and
+      endcolumn = 0
+    }
+
+    private Module getModule() {
+      exists(HTML::ScriptElement script | script.getFile() = file |
+        result.getTopLevel() = script.resolveSource()
+        or
+        not exists(script.getSourcePath()) and
+        // XXX is there no way explicit relation between a script tag and its inline script?
+        result.getTopLevel() instanceof InlineScript and
+        result.getFile() = script.getFile()
+      )
+    }
+
+    override DataFlow::Node getOwnOption(string name) {
+      exists(Module m, DefiniteAbstractValue abstractOptions |
+        any(AnalyzedPropertyWrite write).writes(abstractOptions, name, result) and
+        m = getModule()
+      |
+        exists(ExportDeclaration export, DataFlow::AnalyzedNode exported |
```

Comment:
This needs a few inline comments and ideally some library work to make it easier to implement. The Vue.js model should not be the place to special-case various ways of doing exports in ECMAScript modules.

Reply:
I agree, I will land that in a separate PR (https://jira.semmle.com/browse/ODASA-7747).

Reply:
Can we still have the inline comments for now?

Reply:
I think the rebase has messed with your review. See https://github.com/Semmle/ql/pull/828/commits/cecdd8f2eb61e10c161a456c42f3547b28cf0782#diff-43f0b85ad6846e3bd6a6a50fe0c327a3R339

Or do you mean more inline comments?

Reply:
Yeah, I didn't see those comments when I last looked at it.